### DISCUSSION TITLE
Improvements to the README and package file

### DISCUSSION THREAD

**Discussion by JensenNick** [JensenNick/huawei_solar_pees#10] [Announcements]
Please ONLY use this post to let me know if something could be done to improve the user experience with the existing README and package. Please post "new development" or "alternative solutions" in seperate posts. Thank You!

I've noted down the following ideas / suggestions to improve the understanding, installation process and servicing of the package file.

- [x] Add the repository URL to the package file.
- [x] Add date and/or version numbering.
- [ ] Implementation of the efficiency corrected power sensor in the package file (so you don't need to copy it in at every revision).
- [ ] Setup that allows you to save your settings from revision to revision.
 
Please feel free to contribute with your suggestions for improving the user experience - or solutions to the suggestions above.

**Comment by JacobsenKim** [comment]
> Add date and/or version numbering (not a manual solution).

Jeg har prøvet at kigge på auto tag, men kunne ikke få det til at virke

[packages/git/auto_tag](https://github.com/JacobsenKim/Test/blob/e60c1ca38dddaa328a73ddcc3317abbc2dd8276e/git/auto_tag)

den der ligger under .git/hooks/post-commit får jeg fejl 

> ERROR: arg must be a description of the release in quotes

Jeg har dog ikke styr på hvad der sker da det ikke lige er noget jeg har rodet med før, jeg brugte codespaces her på git til at prøve det af med. 



men den første del 

> How to manage the version number in Git?

beskriver meget godt hvordan man håndtere versions nr.

Jeg vil stadig opfordre til at lave det med et versions nr, selv manuelt til du finder en måde at gøre det auto.
Evt. brug "Releases" til højre i dit "repositorie" også taster du versions nr. ind i tag, så har du et nr. for hvert release og man kan ikke genbruge tags, den der henter din pakke har nu en .zip fil på deres maskine med versions-nr.

Det gør det nemt at rulle tilbage, hvis en ny ændring har lavet problemer, da man har et billede af ens git i det man laver en "Releases"
Se Releases som et øjebliks billede af dit "repositorie" lidt lige som på en VM maskine.
Sker der det at man får lavet et problem så kan der nemt hentes den tidligere pakke og starte derfra.
+ man kan "Set as a pre-release"

Det er meget nemmere hvis der er et verions nr. at referere til ved problemer, også ved bug report.

Det tager ikke 10 sec. at oprette et "Releases"

1)
![Skærmbillede fra 2024-01-29 15-28-53](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/119f15ec-64cf-41b5-abb6-1ecb42ea73ed)

2)
![Skærmbillede fra 2024-01-29 15-25-34](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/fcd24613-555e-48fc-9a53-425c97f92ae3)

3)
![Skærmbillede fra 2024-01-29 15-27-01](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/47690262-16b5-4742-a7db-602b689293b1)

4)
![Skærmbillede fra 2024-01-29 15-27-30](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/ffdaa9fb-9d7d-4a16-8058-8062387da9c7)

5)
![Skærmbillede fra 2024-01-29 15-28-13](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/68ff6138-ebc4-463b-b1ac-0a5abb77c905)

6)
![Skærmbillede fra 2024-01-29 15-40-13](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/97841b62-f1c7-4991-9f4c-a50b102361a8)

**Reply by JensenNick** [reply]
@JacobsenKim Tak kigger på det. Kan det køres sammen med github.dev hvor jeg reviderer package filen?

**Reply by JacobsenKim** [reply]
det virke ikke til der er ret mange script der virker i web versionen.

de script der er omtalt i den fil jeg linker til skal der være en konsol man kan taste i, som i codespace https://github.com/codespaces eller hvis du har installeret VS studio, ved ikke om man kan lave "Releases" fra github.dev.

> The 'tag-version' extension is not available in Visual Studio Code for the Web.

Det er heller ikke meningen at push en update v-nr hver gang der er lavet en ændring i en fil, mere når der er lavet noget nyt, eller der er fixed og testet en eller flere bugs kan man lave en pre release når der er go info tilbage på den kan man fjeren pre-release også er den final, alt i mens du kan arbejde videre som du gør nu, så når du har lavet noget der er teste så kan du push en ny version ud.
Henter man direkte fra filen på siden som nu så får man en "beta" da der er/kan være lavet ændringer efter en "Releases"

**Reply by JacobsenKim** [reply]
@JensenNick 
Jeg fandt et script som kører i codespace (du finder det nederst i den fil jeg linkede til før), det laver et tag hver gang man laver et commit, da opdatere den en package.json med versions nr. som man har i roden af ens repostory og laver så en tag pakke, som man så manuelt skal ud på git siden og lave om til en release ved at klikke på en knap også gøre lidt det jeg viste med billeder, udfylde tekst, men versions nr. (tag nr.) kører auto, du kan se det under mine releases og klikker på tag, så burde du kunne se en masse tags pakker klar til at blive released alle med versions nr.



https://github.com/JacobsenKim/Test/releases

**Reply by JensenNick** [reply]
@JacobsenKim Tak for input. Husk, jeg er autodidakt ud i alt hvad der har med programmering og kodning at gøre. Kan yaml til husbehov, "that's it". Det bliver alt for avanceret til mig og når der så er gået en måned uden at jeg har brugt det, så har jeg glemt alt og skal starte forfra. Jeg har brug for en "set & forget" løsning. Afviser dog ikke "Releases", når du nu har lavet en så god vejledning, som jeg kan vende tilbage til - alternativt blot en versions nr. i filen når nu vi er ovre denne indledende bølge af justeringer. Jeg er ikke begejstret for at der laves zip-filer. Det kan volde problemer for de, der ikke er så PC-kyndige både ift. håndteringen af filen, men også ift. anti-virus programmer.

**Reply by JacobsenKim** [reply]
> Jeg har brug for en "set & forget" løsning,

Det er dette også når det er sat op laver den en tag pakke med nyt nr. hver gang du trykker commit, man kan sætte den om så man manuelt skal gøre det ved at skrive hvilken del af v-nr. man vil opdatere, dog er det ikke nemmere end at gøre som i min vejledning oven for, andet end den ændre nr. i en fil i pakken og at den holder styr på versions nr. som man alligevel kan se i listen når man skal lave nyt tag (som vist ovenfor i vejledningen.) og som du skriver så er du ikke skarp i linux og terminal, så skal du se bort fra en auto løsning da auto = script i baggrunden, dog er der pakker der kan hentes men der er lidt til ingen info om pakker i VS code, så virker det ikke lige ud af boxen som 90% ikke gør kan man kun afinstallere igen, jeg har prøvet flere af de pakker der inde men de gør ikke andet end at installere også se fine ud :-D.

At lave pakker til zip det er sådan git virker, det vil også gøre at du kan køre en latest release version også have pre-release, også har du som nu selve kilde koden, selv om du laver pakker i zip så kan man stadig hente kilde koden som nu, dog henter man så en version som arbejdes på og man må forvente at der kan være problemer, henter man latest release så er man mere sikker på at det er testet, og ønsker man at være på test bænken kan man køre pre-release, på den måde kan du sikre at dem der ikke er så kyndige ikke kommer i problemer med mindre de selv har valgt en pre-release eller hente direkte fra kilde koden.

Så flow vil være kilde kode til pre-release når den er meld ok så ændres den til release.

På den måde kan du sikre at dem der ikke er så kyndige med koder og it, de kan hente en release som er testet uden problemer med et versions nr på filen samt din licens, readme og alt andet er med, de har den fulde pakke på deres PC.
tror det er nemmere at forklare at pakke en fil ud (zip er i alle windows i dag), end at de skal side og rode med nulstilling eller redigering af sensor fordi alle henter fra kilde koden.

How to Open ZIP Files on Windows 10:
https://youtu.be/sruwYBaXvE4?si=dvMp5WWyRYQ8LN3R

Så er det det at du måske får lavet nogle store ændringer som så viser sig at det giver problemer, her kan du nemt hente sidste pre-release eller release alt efter hvordan du vil køre det, og køre tilbage (starte forfra) det kan spare en mase tid, eks. kan være med float(0) hvor du mistænker den for at give en masse problemer her kunne du så hente sidste pakke og lave de få ændringer du havde lavet før float(0) og du er hurtigt tilbage på sporet, samt alle der har valg ikke at køre pre-release ikke ville være berørt da de stadig kører på den stabile release.

Men ja et versions nr. øverst i filen vil være en god start, ja man vil glemme at rette nr. til det bliver en vane, og selv der vil man glemme at rette en gang i mellem men det er bedre end nu hvor man kører ren kilde kode uden hjelm på og alle har en forskellig version alt efter hvornår de har hentet den og ingen ved hvilket version de har.

**Reply by JacobsenKim** [reply]
Fandt lige ud af selv om man har lavet en zip pakke (release, tag, pre-release) så kan man stadig tilgå filerne fra git siden

![Skærmbillede fra 2024-01-31 06-00-12](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/1ae49bf0-029a-46c5-b459-7ff1e50da273)


Her er filerne i tag 11 og via menu kan man vælge mellem alle tag

![Skærmbillede fra 2024-01-31 06-02-05](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/ef3d5f43-fe89-4826-b158-ff02d8ea36c0)

**Comment by JensenNick** [comment]
@JacobsenKim Jeg forsøger at lave en release, men går i stå hvor jeg skal til at vedhæfte filer / etablere zip-filen. Jeg kan kun se muligheder for at vedhæfte filer fra min PC. Det må være muligt at vedhæfte det der ligger på GitHub? Har også gemt en draft (kladde) af "Release" som jeg er i gang med og hvor mon den kan findes?

**Reply by JacobsenKim** [reply]
det gør den selv, du skal ikke lave zip filerne

Det er kun hvis du har ydeliger filer du vil have med, mest brugt ved programmer hvor de er lavet for en, eks. kan være en installations fil til windows, men det har du ikke noget af.

**Reply by JacobsenKim** [reply]
Det er så nemt som på guiden klik klik skriv din info klik

**Reply by JacobsenKim** [reply]
Din draft finder du under release, klikke på rediger også udgive den som guiden ovenfor.

This is a draft and won’t be seen by the public unless it’s published.

Du kan altid slette en release igen, du kan også ændre din tekst og billeder under en release.

**Reply by JacobsenKim** [reply]
Lav evt. en ny category i din discussions med navnet Releases

![Skærmbillede fra 2024-02-01 18-52-28](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/1fe2fe21-1c3d-4f41-8bce-dcbbf09c8184)



Så kan du lave en post auto fra release siden med et flue ben også vælge release category i din discussion.




![Skærmbillede fra 2024-02-01 18-56-44](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/aa13389e-93e1-431f-95a2-2443f394887e)

**Comment by JacobsenKim** [comment]
Info ang. 1 Inverter:

Jeg har kørt v1.0.0 ind på en VM med en backup af min HA før PEES (så der er ikke nogen spor af PEES på den) og roder lidt med at få 1 inverter til at køre.

sensor.power_inverter_2_input ser ud til at kører med 0 fra start.

sensor.energy_yield_2 kan man godt sætte til 0 som beskrevet også virke det, men den re-setter efter genstart.

Men sætter man sensor.energy_yield_2_ps set to 0, så virker det også med genstart, som det ser ud nu.

Men nu skal jeg lige rode lidt mere så re-setter jeg VM og laver en frisk install af pees, hvor jeg så prøver kun at sætte sensor.energy_yield_2_ps til 0, for lige at teste det af at det er løsningen på en helt ren install, tror ikke jeg når det i dag før solen er helt væk så det bliver nok i morgen jeg får prøvet med en frisk install.

**Comment by JacobsenKim** [comment]
ang. https://github.com/JensenNick/huawei_solar_pees/wiki/FAQ-%E2%80%90-Frequently-Asked-Questions

Anbefal at laver en backup inden man går i gang med at slette løs. :-D

**Comment by JacobsenKim** [comment]
Så fik jeg kørt VM tilbage og kørt PEES ind igen, denne gang kun ved at sætte `sensor.energy_yield_2_ps set to 0` så kommer `sensor.energy_yield_2 ` auto til 0, jeg søgte på sensor.energy_yield_2 så kommer begge frem og man kan se at begge går i nul, denne måde virker også ved genstart, jeg har så testet videre da jeg ikke kunne vente på at alle mine sensor blev aktive, så dem har jeg også sat deres `_ps` til 0, på denne måde kommer regnskabet hurtiger igang end der er en sensor som for mig i dette eksempel først bliver aktiv i morgen tidligt, så det kan anbefales at sætte de sensor til 0 for at få tallene til at passe sammen.


![Skærmbillede fra 2024-02-02 18-17-14](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/307c6827-a386-4f4e-913a-0b56dab7ba3b)


Her har jeg ladet den køre lidt og samle data før jeg har genstartet og alt er stadig ok.


![Skærmbillede fra 2024-02-02 18-39-55](https://github.com/JensenNick/huawei_solar_pees/assets/157890151/f933ef15-4318-4e52-9335-9a078b124238)

**Reply by JensenNick** [reply]
@JacobsenKim Tak for input. Jeg tænker om det kunne løses på en bedre måde ved at angive en `last = 0`. Den skal selvfølgeligt konfigureres således, at den overskrives lige så snart at der optræder en målt værdi. Tror også at dette vil løse det jeg har kommenteret i v1.0.0 tråden. Men, jeg har for mange jern i ilden lige nu...

**Reply by JacobsenKim** [reply]
Måske noget ala dette, du kan prøve det af i skabelon.

Indsæt en sensor du ved findes og ved laver et tal
Jeg har 2 med samme navn på nær 01 og 02 så det var nemt at skifte mellem

Denne findes ikke: sensor.luftkvalitet_0_temperature
Denne findes men slukket: sensor.luftkvalitet_01_temperature
Denne findes og tændt: sensor.luftkvalitet_02_temperature


Sensor findes og tændt: {{ states('sensor.luftkvalitet_02_temperature') }}
state: >
  {% set x = states('sensor.luftkvalitet_02_temperature') %}
  {{ x if (x != 'unavailable' and x != 'unknown' and x is not none) else 0.0 }}
  
  Nu kan du skifte sensor ud med en der ikke finde og en der ikke er tændt, begge skal vise 0.0 hvor den der er tændt og aktiv vil vise dens tal.
  


Her en anden måde at gøre det på
Sensor findes og tændt: {{ states('sensor.luftkvalitet_02_temperature') }}
      value_template: >-
        {% if states('sensor.luftkvalitet_02_temperature') not in ['unknown', 'unavailable'] %}
          {{ states('sensor.luftkvalitet_02_temperature')|round(2) }}
        {% else %}
          {{ ( states ('0.0'))|float(0)|round(2) }}
        {% endif %}

OBS. Det er _ps der skal sættes, så kører alt andet.

**Reply by JensenNick** [reply]
@JacobsenKim Jeg er temmelig sikker på at disse forslag vil medføre at sensorerne går i nul når de bliver unknown / unavailable. Problemet med dette er at sensorværdier og utility meters springer voldsomt i værdi. Lidt afhængigt af hvilken state_class de har. En af de største udfordringer jeg har haft, har været at skabe noget det ikke laver disse ukontrollerede spring ved udfald / genstart.

**Reply by JacobsenKim** [reply]
Ja det er lige hvordan udfald bliver, ellers gå med at man ved installation skal sætte de _ps sensor der ikke har en nummer værdi til 0
lige efter første genstart, så ved du det kører, igen det er kun ved installation man skal gøre det og alle sensor kommer frem når man søger på _ps så det tager max. 1 min. at sætte de manglende til 0

Men hvis du har en ide til at få det til at virke med koden har jeg stadig min HA jeg kan køre tilbage til før PEES og lave en ren install om og om igen.

Lav evt. en ny branch med navn "start" så kan du ændre i den + jeg kan hente fra den og teste på min VM, finder du løsningen kan du merge den ind i main.

**Reply by JensenNick** [reply]
@JacobsenKim Jeg kan se at man få "det hele" med når man downloader vha. release. Det forekommer uhensigtsmæssigt at billeder følger med, men det skal de måske for at der i README kan refereres til dem hvis / når README åbnes på lokal maskine?

**Reply by JacobsenKim** [reply]
Ja en release er hele din main.

kigger man i readme.md så referere den billederne til pictures/
I dette eksempel til pictures/flows_definitions.jpg

 Definitions and Flows (pictures/flows_definitions.jpg)
 
 
De ændringer du laver i din main nu kommer med i næste release.

**Reply by JacobsenKim** [reply]
Se hvordan du slipper for at have Wiki billeder i din git pakke
[Billeder gemt i Wiki](https://github.com/JacobsenKim/Test/wiki/Test#billeder-i-wiki)
